src/irmin/contents.ml

```diff
@@ -32,6 +32,222 @@ module Cstruct = struct
+
+let merge_json = Merge.(v json merge_value)
+
+module Json = struct
+  type t = (string * json) list
+  let t = Type.(list (pair string json))
+  let merge = Merge.(option (alist Type.string json (fun _key -> Merge.option merge_json)))
+  let lexeme e x = ignore (Jsonm.encode e (`Lexeme x))
+
+  let rec encode_json e = function
+    | `Null -> lexeme e `Null
+    | `Bool b -> lexeme e (`Bool b)
+    | `String s -> lexeme e (`String s)
+    | `Float f -> lexeme e (`Float f)
+    | `A a ->
+        lexeme e `As;
+        List.iter (encode_json e) a;
+        lexeme e `Ae;
+    | `O o ->
+        lexeme e `Os;
+        List.iter (fun (k, v) ->
+          lexeme e (`Name k);
+          encode_json e v
+        ) o;
+        lexeme e `Oe
+
+  let pp fmt x =
+    let buffer = Buffer.create 32 in
+    let encoder = Jsonm.encoder (`Buffer buffer) in
+    encode_json encoder (`O x);
+    ignore @@ Jsonm.encode encoder `End;
+    let s = Buffer.contents buffer in
+    Fmt.pf fmt "%s" s
+
+  let decode_json d =
+    let decode d = match Jsonm.decode d with
+      | `Lexeme l -> l
+      | `Error e -> failwith (Fmt.strf "%a" Jsonm.pp_error e)
+      | _ -> failwith "invalid JSON encoding"
+    in
+    let rec unwrap v d = match v with
+      | `Os -> obj [] d
+      | `As -> arr [] d
+      | `Null | `Bool _ | `String _ | `Float _ as v -> v
+      | _ -> failwith "invalid JSON value"
+    and arr vs d =
+      match decode d with
+      | `Ae -> `A (List.rev vs)
+      | v ->
+          let v = unwrap v d in
+          arr (v::vs) d
+    and obj ms d =
+      match decode d with
+      | `Oe -> `O (List.rev ms)
+      | `Name k ->
+          let v = unwrap (decode d) d in
+          obj ((k, v) :: ms) d
+      | _ -> failwith "invalid JSON object"
+    in
+    try
+      Ok (unwrap (decode d) d)
+    with
+      | Failure msg -> Error (`Msg msg)
+
+  let of_string s =
+    let decoder = Jsonm.decoder (`String s) in
+    match decode_json decoder with
+    | Ok (`O obj) -> Ok obj
+    | Ok _ -> Error (`Msg "Irmin JSON values must be objects")
+    | Error _ as err -> err
+end
+
+module Json_value = struct
+  type t = json
+  let t = json
+  let merge = Merge.(option merge_json)
+
+  let pp fmt x =
+    let buffer = Buffer.create 32 in
+    let encoder = Jsonm.encoder (`Buffer buffer) in
+    Json.encode_json encoder x;
+    ignore @@ Jsonm.encode encoder `End;
+    let s = Buffer.contents buffer in
+    Fmt.pf fmt "%s" s
+
+  let of_string s =
+    let decoder = Jsonm.decoder (`String s) in
+    match Json.decode_json decoder with
+    | Ok obj -> Ok obj
+    | Error _ as err -> err
+end
+
+module Json_tree(P: S.PATH)(M: S.METADATA) = struct
+  include Json_value
+
+  module type STORE = S.STORE with
+    type step = P.step
+    and type key = P.t
+    and type contents = json
+    and type metadata = M.t
+
+  type tree = [
+    | `Tree of (P.step * tree) list
+    | `Contents of json * M.t
+  ]
+
+  let rec to_concrete_tree (j: json): tree =
```

Comment:
The function is not tail recursive, this could be a problem on large JSON docs (the ones that you want to project to a tree).